### PR TITLE
Feature/au 191 virheilmoitukset

### PR DESCRIPTION
**Pull request by jiisuominen**
# [AU-191](https://helsinkisolutionoffice.atlassian.net/browse/AU-191)

## What was done

* Styled error message shown under title
* Styled fields that have errors

## How to install

* Make sure your instance is up and running on correct branch.
  * `git checkout feature/AU-191-virheilmoitukset`
  * `make fresh`
* Run `make drush-cr`

## How to test

* [ ] Login and go start a new application https://hel-fi-drupal-grant-applications.docker.so/fi/tietoa-avustuksista/avustukset/kaupunginhallitus-yleisavustus
* [ ] First check that you can't see red error field under title, because you do not have any errors yet
* [ ] Next click any other tab and check that error notification looks like this:
image
* [ ] Go back to first page and check that fields that have errors have red line
image

[AU-191]: https://helsinkisolutionoffice.atlassian.net/browse/AU-191?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php [hunk 1]
```diff
@@ -502,6 +502,37 @@ public function alterForm(array &$form, FormStateInterface $form_state, WebformS
...
+        $errorName = strtok($errorKey, ']');
+        $errorSelectValue = substr($errorKey, strpos($errorKey, '[') + 1);
+        if (isset($form['elements'][$pageName][$errorName])) {
+          $form['elements'][$pageName][$errorName]['#attributes']['class'][] = 'has-error';
+        }
+        else {
+          foreach ((array) $form['elements'][$pageName] as $fieldName => $element) {
+            if (!str_starts_with($fieldName, '#')) {
+              if (isset($form['elements'][$pageName][$fieldName][$errorName]['#webform_composite_elements'][$errorSelectValue])) {
+                $storage = $form_state->getStorage();
```
Comment (jiisuominen) [inline public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php:523]: This should be defined outside of foreach.

This way you'll be getting that storage as many times as the foreach is run.

**Review by jiisuominen — COMMENTED** [review]
Please use comments.

File: public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php (hunk 1 above)
Comment (jiisuominen) [reply public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php:523]: Storage variable I mean.

**Comment by jiisuominen** [comment]
`Warning: Undefined array key "webform_preview" in Drupal\grants_handler\Plugin\WebformHandler\GrantsHandler->alterForm() (line 520 of modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php).
Warning: Undefined array key "webform_confirmation" in Drupal\grants_handler\Plugin\WebformHandler\GrantsHandler->alterForm() (line 520 of modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php).`

These should also be fixed somehow.

File: public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php [hunk 2]
```diff
@@ -502,6 +502,37 @@ public function alterForm(array &$form, FormStateInterface $form_state, WebformS
         ->addError('Application period is closed, no further editing is allowed.');
       $form['#disabled'] = TRUE;
     }
+
+    $all_current_errors = $this->grantsFormNavigationHelper->getAllErrors($webform_submission);
+    $storage = $form_state->getStorage();
```
Comment (jiisuominen) [inline public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php:507]: Ok, so the $storage could be used better here, to set our form errors to a subarray:

`
$storage = $form_state->getStorage();
$errors = $storage['errors']
...
 `

File: public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php [hunk 3]
```diff
@@ -502,6 +502,37 @@ public function alterForm(array &$form, FormStateInterface $form_state, WebformS
...
+        $errorSelectValue = substr($errorKey, strpos($errorKey, '[') + 1);
+        if (isset($form['elements'][$pageName][$errorName])) {
+          $form['elements'][$pageName][$errorName]['#attributes']['class'][] = 'has-error';
+        }
+        else {
+          foreach ($form['elements'][$pageName] as $fieldName => $element) {
+            if (!str_starts_with($fieldName, '#')) {
+              if (isset($form['elements'][$pageName][$fieldName][$errorName]['#webform_composite_elements'][$errorSelectValue])) {
+                $storage[$errorName] = 'has-errors';
+                $form_state->setStorage($storage);
```
Comment (jiisuominen) [inline public/modules/custom/grants_handler/src/Plugin/WebformHandler/GrantsHandler.php:525]: Also, same thing with setting the storage than with getting it. No need to set it many times. Just fill the new $errors array with them and then set that to form state AFTER all foreaches.

**Review by suvikankare — APPROVED** [review]

**Pull request merged**